### ISSUE TITLE
Behaviour cooldown: useful or harmful?

### ISSUE THREAD

**Issue by fabio-t** [fabioticconi/alone-the-roguelite#8]
Right now, at every AI tick (variable for each entity, to avoid clogging the system) we evaluate all possible behaviours and choose the best one. Depending on the circumstances, the scores might be so that we could be in a situation where behaviour A is selected, then behaviour B, then behaviour A, every tick.. thus doing close to no work.

One solution is to add a behaviour cooldown, optional and to be used sparingly. Important, substained behaviours (like Chase and Flee) should have a cooldown lasting twice or thrice a a single movement time. This allows for the same behaviour to be evaluated at each tick, unless its score goes to 0 or the cooldown expires; then the other behaviours would be evaluated again and the best executed.

One problem is of course reactivity: if I put a cooldown to non-essential behaviours, or too long a cooldown in general, the creature will be complitely oblivious to things unrelated to the active behaviour.

With some discipline in cooldown configuration, it should be fine though: behaviours related to escaping death/danger should have the highest cooldown (we don't care about anything else); behaviours related to nutrition should have a cooldown (it's important we don't starve) but proportional to hunger, so for a light hunger we'd eat, but we'd still be highly reactive to the environment; other behaviours don't need a cooldown.

**Comment by clarkaelliott** [comment]
I'd imagine a cooldown only to prevent stacking behaviors in a single tick would be enough to begin testing with?  We don't want unresponsive entities, but at the same time we don't want confused ones!  There has to be a nice middle ground somewhere.

Also, I had no idea you were so active on this project!  Unfortunately I am spread thin at this time, focused on still learning Java fundamentals, I am working my way through my textbooks very thoroughly and building those projects & the same with Android SDK each day.

In short, I currently will only have maybe 5-10 hours a week to look into this project, it will primarily be a learning experience for me, in both game development and the Java language!  Thank you for inviting me in, and sharing all these communications with me, sorry I cannot provide more help at this time, but I am working on it each day :+1:

**Comment by fabio-t** [comment]
I open issues for myself really, so I don't get too distracted at work (I brain-dump stuff here and forget it, then come back to it at night).

You are free to look around, play with things etc, as github allows - you would be working on your forked version, not on my repository :) so you can do/undo as you wish, for learning.

**Comment by fabio-t** [comment]
Regarding your question:

> I'd imagine a cooldown only to prevent stacking behaviors in a single tick would be enough to begin testing with? We don't want unresponsive entities, but at the same time we don't want confused ones! There has to be a nice middle ground somewhere.

There is always a single active behaviour at any time, never more. The tick is also flexible, by entity. So, say entity 1 is grazing peacefully on the grass, and a predator arrives. At the next AI tick for entity 1, we'd evaluate all behaviours: FleeBehaviour would spot a predator and probably be selected and executed. That means that entity 1 would be assigned a MoveTo component to be processed by the MovementSystem at the right time.

However, it may happen that the predator stays far away, and entity 1 Hunger at the next tick is grown enough that it "wins" against the FleeBehaviour. So we switch to GrazeBehaviour again. Then the predator comes nearer, and we switch to FleeBehaviour again, moving a single cell.

This might be ok, but in general I find it non optimal. I think when you enter "FleeBehaviour" you should move at least 2-3 cells away from the predator before even looking if you are hungry or if your group is far away (that would be FlockingBehaviour).

Kind of an "adrenaline rush": if you are in a dangerous situation you focus on fleeing, even if the predator has actually stayed behind (maybe it's tired/hurt, or maybe it's chasing another prey).

It shouldn't actually be needed, this cooldown, if the utility functions of all behaviours are well coordinated. But it's hard to check if they are.
